core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -754,4 +837,23 @@ public static StreamInput wrap(byte[] bytes) {
     public static StreamInput wrap(byte[] bytes, int offset, int length) {
         return new InputStreamStreamInput(new ByteArrayInputStream(bytes, offset, length));
     }
+
+    /**
+     * A functional interface to allow functions to apply with a checked exception.
+     * <p>
+     * This is identical to {@link java.util.function.Function}, except that it also allows for an {@code IOException} to be thrown,
+     * and it only allows {@link StreamInput} to be used as an argument.
+     */
+    @FunctionalInterface
+    public interface StreamInputFunction<R> {
```

Comment:
As I was removing my added methods, I noticed that apparently `Writeable.Reader<R>` also adds this one (just FYI).